render.gotmpl

```diff
@@ -13,8 +21,39 @@ import (
 type {{ $applyOptionFuncType }} func(c *{{ $.configTypeName }}) error
 
 func (f {{ $applyOptionFuncType }}) apply(c *{{ $.configTypeName }}) error {
-return f(c)
+    return f(c)
+}
+
+{{ $decoratedApplyOptionFuncType := printf "Decorated%s" $applyOptionFuncType }}
+
+type {{ $decoratedApplyOptionFuncType }} struct {
+	name string
+	args []interface{}
+	{{ $applyOptionFuncType }}
+}
+
+func Decorate{{ $applyOptionFuncType }}(name string, args []interface{}, f {{ $applyOptionFuncType }}) {{ $decoratedApplyOptionFuncType }} {
+	return {{ $decoratedApplyOptionFuncType }}{
+{{ if ( or .implementStringer .implementEqual ) -}}
+		name: name,
+		args: args,
+{{ end }}
+		{{ $applyOptionFuncType }}: f,
+	}
+}
+
+{{ if .implementStringer -}}
+func (f {{ $decoratedApplyOptionFuncType }}) String() string {
+	return fmt.Sprintf("%s: %v", f.name, f.args)
 }
+{{ end }}
+
+{{ if .implementEqual -}}
+func (f {{ $decoratedApplyOptionFuncType }}) Equal(o interface{}, opts... cmp.Option) bool {
```

Comment:
Should we be concerned about exposing cmp as part of our api?  cmp comes with a caveat that it's meant only for testing.